### ISSUE TITLE
Does it make sense to separate getUser() and getToken() methods from context?

### ISSUE THREAD

**Issue by Him2021** [HNeukermans/adal-ts#10]
I will explain what I mean. I have implemented adal-ts in my project. A couple of scenarios in my project (and I think these are very regular scenarios)  are to display UserName across the site and implement auth guards. 

- The username code has been implemented in my app component. I have to create a context here with Create config code to do _context.getUser();_

- In my auth guard, I am checking if _context.getuser()_ returns null. So I have create config again.

- To call my web api, I need to pass the token and need to create config again. 
Now, I can create a service and inject context but looking at adal-ts code it looks like it doesnt need context object to get user or token. 
Does it make sense to refactor the code so that I dont have to create config object again and again?

**Comment by HNeukermans** [comment]
indeed @Him2021. Good feedback. 
I already proposed a similar idea in previous issues.  [link](https://github.com/HNeukermans/adal-ts/pull/9)

To retrieve the token or user y**ou only need to talk to the storage**. That's why I proposed a change to the api:  (repeat [link](https://github.com/HNeukermans/adal-ts/pull/9))

For getting the user, instead of writing 
- Authentication.getContext().getUser()

**better would be**  
- new User(Authentication.getStorage().getIdToken()) ||
- User.parse(Authentication.getStorage().getIdToken()) ||
- Authentication.getStorage().getIdToken().parse()

Which one do you prefer?  Feel like implementing ?

**Comment by Him2021** [comment]
I think the last option would be better as it goes with the rest of the methods for login and logout and will ensure consistency. I will try and contribute but the next few days are booked solid for me with my project deadlines.

**Comment by frederikprijck** [comment]
My two cents:

I don't realy like any of the proposed api's:

- new User(Authentication.getStorage().getIdToken())
- User.parse(Authentication.getStorage().getIdToken())
- Authentication.getStorage().getIdToken().parse()

Wouldn't it be a better idea to access those using an API like

- Authentication.getUser()
- Authentication.getToken()

Example:
```
export class Auth {
    constructor(
        private localStorage: LocalStorage,
        private userDecoder: UserDecoder
    ) {
    }

    public getUser(): User {
        const idToken = this.getToken();

        try {
            let user = this.userDecoder.decode(idToken);
            return user;
        } catch (error) {
            if(console && console.debug) console.debug('getUser() returns null on catched error. Details >> '+ error.toString())
            return null;
        }
    }

    public getToken(): string {
        return this.localStorage.getItem(Constants.STORAGE.IDTOKEN)
    }
    
    public getContext(configuration: AdalConfig): AuthenticationContext {
        return new AuthenticationContext(
            configuration,
            this.localStorage,
            new Navigator(),
            new GuidGenerator(),
            new AadUrlBuilder(new GuidGenerator()),
            this.userDecoder,
            new AadLogoutUrlBuilder());
    }

    public getAadRedirectProcessor() {
        return new AadRedirectProcessor(
            new QueryStringDeserializer(),
            this.userDecoder,
            this.localStorage,
            window);
    }
}

export const Authentication = new Auth(new LocalStorage(), new UserDecoder());
```

**Comment by HNeukermans** [comment]
To understand, the reasoning behind this api and why I came up with this:

- `Authentication.getStorage().getIdToken()`

we must enter a broader discussion. It is a discussion about making the behavior of the components configurable.

Currently the components behavior, like, for example, the storage - and navigator components, is fixed and can not be changed by the api user. In future releases I would like the api user to allow  the behavior of the components to be configurable. Dependency injection can make this technically possible.... and the class best fitted for serving the components would be the Authentication class, in my idea. (**)

lets say the end user prefers:
- to use sessionstorage instead of localstorage.
- the end user wishes to show the Azure AD login box inside a popup, rather to navigating to a separate page. 

```
AdalBuilder.setStorage(storage: Storage): void;
AdalBuilder.setNavigator(navigator: Navigator): void;
let authentication: Authentication = AdalBuilder.build();
```
This way the end user can provide his own implementation of storage or navigator or chose from the one that come with adal-ts. 

inside adal for example we could provide a standard PopupNavigator and PageNavigator and a standard 
SessionStorage and LocalStorage.
```
class PopupNavigator implements Navigator
class Pagenavigator implements Navigator

class SessionStorage implements Storage
class LocalStorage implements Storage
```

This answer is far from complete, but it should give you an idea, of where i would like to go with adal-ts.  
Offcourse I am always open for discussions or ideas about this topic.

As for now: by what api the end user retrieves the token or user, is really not that important. 
What is most importnant now, for the end user is that both token and user retrieval are supported by adal-ts. 

(**) since user and token are not building block (= components with configurable behavior) I would not like to see Authentication.getUser() en Authentication.getToken().

**Comment by frederikprijck** [comment]
I fully agree looking at it from that perspective, getUser and getToken shouldn't be part of Authentication itself.

**Comment by HNeukermans** [comment]
You'r welcome. 
Thx for the contributions so far.
